src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotUseEndOfStreamInAsyncMethods.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    using static MicrosoftNetCoreAnalyzersResources;
+
+    /// <summary>
+    /// CA2024: <inheritdoc cref="DoNotUseEndOfStreamInAsyncMethodsTitle"/>
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class DoNotUseEndOfStreamInAsyncMethodsAnalyzer : DiagnosticAnalyzer
+    {
+        private const string RuleId = "CA2024";
+        private const string EndOfStream = nameof(EndOfStream);
+
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(DoNotUseEndOfStreamInAsyncMethodsTitle)),
+            CreateLocalizableResourceString(nameof(DoNotUseEndOfStreamInAsyncMethodsMessage)),
+            DiagnosticCategory.Reliability,
+            RuleLevel.BuildWarning,
+            CreateLocalizableResourceString(nameof(DoNotUseEndOfStreamInAsyncMethodsDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var streamReaderType = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIOStreamReader);
+
+            if (streamReaderType is null)
+            {
+                return;
+            }
+
+            var endOfStreamProperty = streamReaderType.GetMembers(EndOfStream)
+                .OfType<IPropertySymbol>()
+                .FirstOrDefault();
+
+            if (endOfStreamProperty is null)
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(AnalyzePropertyReference, OperationKind.PropertyReference);
+
+            void AnalyzePropertyReference(OperationAnalysisContext context)
+            {
+                var operation = (IPropertyReferenceOperation)context.Operation;
+
+                if (!IsAsyncContext(context))
+                {
+                    return;
+                }
+
+                if (SymbolEqualityComparer.Default.Equals(streamReaderType, operation.Member.ContainingType) &&
```

Comment:
Which is cheaper, checking whether the current method is async or checking whether operation is the EndOfStream symbol... I assume the latter? If yes, we should do that as the first thing so that we don't waste more time examining the method. Use of EndOfStream is relatively rare, while use of properties in async methods is very common.

Reply:
Good catch, I compared the performance with AnalyzerRunner from Roslyn (tested on the ASP.NET Core repository, no log file, single-threaded, no Windows Defender):

**Original**
```
Execution times (ms):
DoNotUseEndOfStreamInAsyncMethodsAnalyzer:     367
DoNotUseEndOfStreamInAsyncMethodsAnalyzer:     351
DoNotUseEndOfStreamInAsyncMethodsAnalyzer:     353
```

**With checking symbols first**
```
Execution times (ms):
DoNotUseEndOfStreamInAsyncMethodsAnalyzer:     337
DoNotUseEndOfStreamInAsyncMethodsAnalyzer:     334
DoNotUseEndOfStreamInAsyncMethodsAnalyzer:     338
```

Checking symbols first is about 5% faster in this example (and intuitively should also be faster in general based on your reasoning), so this seems like a good change.